### DISCUSSION TITLE
Can I invoke/execute a macro conditionally during Startup process?

### DISCUSSION THREAD

**Discussion by pimgeek** [TiddlyWiki/TiddlyWiki5#6052] [Q&A] [answered]
I tried to use various methods( $list, $set, $reveal ), but all of them are failed ...

The follwing Tiddler had been tagged with `$:/tags/StartupAction/PostRender`

```
<$list variable="action" filter="[[$:/status/IsReadOnly]get[text]search-replace[yes],[hide-elements]]">
<$macrocall $name=<<action>> />
</$list>
```
Then I tried to use JS Macro, without success, either

```JavaScript
(function(){

/*jslint node: true, browser: true */
/*global $tw: false */

"use strict";

exports.name = "reader-mode";

exports.params = [];
/*
Run the macro
*/
exports.run = function(act) {
  var isReadOnly = ""
  var statusTid = this.wiki.getTiddler("$:/status/IsReadOnly")
  if (statusTid) {
    isReadOnly = statusTid.getFieldString("text")
    if (isReadOnly == "yes") {
      return "readonly-hide-elements"
    }
  }
  return "unhide-elements"
};
})();
```

The follwing Tiddler had been tagged with `$:/tags/StartupAction/PostRender`

```TiddlyWiki
\import [[$:/core/ui/PageMacros]] [all[shadows+tiddlers]tag[$:/tags/Macro]!has[draft.of]]
\import $:/plugins/kookma/utility/author/tools

<<reader-mode>>

<$macrocall $name=<<reader-mode>> />
```

**Comment by pmario** [comment]
Hi,
Have a closer look at: https://tiddlywiki.com/#StartupActions

**Reply by pimgeek** [reply]
I have read this and written some simple js macros success fully, but that's much simpler date calc task -- https://wiki.hintsnet.com/#%2Fjsmacro%2Fday-diff.js:%24%3A%2FAdvancedSearch%20%2Fjsmacro%2Fday-diff.js%20%24%3A%2Fpimgeek%2Fstartup

**Comment by pmario** [comment]
JS macros have to have a header, that is needed by the startup mechanism, otherwise the macro won't be available

```
/*\
title: $:/plugin/your name/macros/readermode.js
type: application/javascript
module-type: macro

Your description of the macro 3-5 lines

\*/

/* your code comes here*/
```

**Reply by pmario** [reply]
>Then I tried to use JS Macro, without success, either

How did you create the macro?

**Reply by pimgeek** [reply]
Would you have a look at this website？I want to hide some ui elements for anonymous visitors

https://wiki.hintsnet.com/#%24%3A%2Fpimgeek%2Fstartup:%24%3A%2FAdvancedSearch%20%24%3A%2Fpimgeek%2Fstartup

**Reply by pmario** [reply]
The tiddler `$:/status/IsReadOnly` is defined by the syncer, when it asks the server for the status info. 
This communication is done after the rendering has already happened. 

Have a closer look at the Network tab in the Dev-console. 

![grafik](https://user-images.githubusercontent.com/374655/133931815-bfeee571-0909-4b14-932a-0fc0b89ca7b3.png)

As you can see. The tiddler doesn't exist at 1 Since it is created at 3

**Reply by pmario** [reply]
You may try to use the tiddler `$:/status/IsReadOnly` directly .. There may be some blinking, but it should work.

**Reply by pimgeek** [reply]
Thanks! Now I have a new plan, make every reader-related UI elements reveal / conceal according to the content of `$:/status/IsReadOnly`. ☺

**Comment by pmario** [comment, marked as answer]
So your js-macro does work, but the readOnly status tiddler doesn't exist, when it is executed.

**Reply by pimgeek** [reply]
This time my solution works!

```JavaScript
/*\
title: $:/pimgeek/reader-mode.js
type: application/javascript
module-type: macro

\*/
(function(){

/*jslint node: true, browser: true */
/*global $tw: false */
"use strict";

/*
Information about this macro
*/
exports.name = "reader-mode";

exports.params = [];

/*
Run the macro
*/
exports.run = function() {
  function fb(err, isLoggedIn, username, isReadOnly, isAnonymous) {
    if (err){
      $tw.utils.warning(err)
    } else {
      $tw.wiki.addTiddler({ title: "$:/status/IsReadOnly", text: isReadOnly ? "yes": "no", "hidden": "$:/ControlPanel $:/core/ui/SideBar/More" })
    }
  }
  $tw.syncadaptor.getStatus(fb)
};

})();
```

Changes to Sidebar Template Tiddler

`<$list filter="[all[shadows+tiddlers]tag[$:/tags/SideBarSegment]!has[draft.of]]` **-[[$:/status/IsReadOnly]text[yes]get[hidden]split[ ]]** `" variable="listItem">`

StartupAction/PostRender Tiddler
```
<<reader-mode>>
```
